### PR TITLE
Add error_state to all config dump resources

### PR DESCRIPTION
**Pull request by lidizheng**
Signed-off-by: Lidi Zheng <lidiz@google.com>

Commit Message: Add error_state to all config dump resources
Additional Description: Store the NACKed resource in each resources
Risk Level: None
Testing: None
Docs Changes: With proto definition
Release Notes: Add error_state to all config dump resources
Platform Specific Features: None
Fixes: https://github.com/envoyproxy/envoy/issues/14431

### CODE REVIEW

File: api/envoy/admin/v2alpha/config_dump.proto
Comment (markdroth) [inline api/envoy/admin/v2alpha/config_dump.proto:52]: There's no need to make changes in v2alpha at this point.  Let's just make the change in v3.

File: api/envoy/admin/v3/config_dump.proto
Comment (markdroth) [inline api/envoy/admin/v3/config_dump.proto:192]: Just to check my understanding here, this is how I think this will work:

If we are currently using version 1 of a resource because we NACKed version 2, then `DynamicCluster.version_info` will be "1", `DynamicCluster.cluster` will be the contents of version 1, and `DynamicCluster.error_state.version_info` will be "2".

Is that right?  We should probably add comments to all of these fields clarifying exactly what they will contain.

Note that the semantics of the individual fields may be different when the CSDS output comes from an xDS client rather than an xDS server, and these protos are used in both contexts, so the comments should make that clear.  The summary I gave above is for an xDS client only.

**Review by markdroth — COMMENTED** [review]
Thanks for doing this, Lidi.

File: api/envoy/admin/v3/config_dump.proto
Comment (lidizheng) [reply api/envoy/admin/v3/config_dump.proto:198]: Comment updated.

Yes, the example you described is exactly how this field gonna work. Also, I think the same mechanism works for xDS servers as well. CSDS is used by xDS servers to answer a specific client's configuration status, so it's valid to have the config dumped resource version match the in-effective version and have the rejected details stored separately. WDYT?

CC @fuqianggao

**Review by lidizheng — COMMENTED** [review]
No problem! Thanks for the quick review.

File: api/envoy/admin/v3/config_dump.proto
Comment (srini100) [inline api/envoy/admin/v3/config_dump.proto:140]: *particular (fix other places too)

Comment (lidizheng) [reply api/envoy/admin/v3/config_dump.proto]: Good catch! Fixed.

File: api/envoy/admin/v3/config_dump.proto
Comment (markdroth) [reply api/envoy/admin/v3/config_dump.proto:198]: We should also add a comment to the `UpdateFailureState.failed_configuration` field that it may not be populated on clients.  If we've NACKed the most recent version, we can record that version, but I don't think we want to record the entire failed resource.

Comment (lidizheng) [reply api/envoy/admin/v3/config_dump.proto:200]: Updated the comment, basically saying this field is optional.

File: api/envoy/admin/v3/config_dump.proto
Comment (markdroth) [inline api/envoy/admin/v3/config_dump.proto:53]: I don't think this is going to be an issue on the server.  The server will already store the most recent version, because that's what it's expecting to serve.

In contrast, the server may not want to populate the version currently being used by the client if it has NACKed the most recent update, because that would require the server to hold on to an older version of the resource.

Comment (lidizheng) [reply api/envoy/admin/v3/config_dump.proto]: Comment updated to detail description of the expectation of xDS client and xDS server.

Comment (markdroth) [reply api/envoy/admin/v3/config_dump.proto]: The comment about the server isn't relevant to this field, though.  If the client NACKs, then this field is always the most recent version.

I think this comment should only mention the client.  The comment about the server is probably more relevant in the fields that contain the version that the client is currently using.

Comment (lidizheng) [reply api/envoy/admin/v3/config_dump.proto]: Updated the comment to only mention the xDS client.

**Comment by markdroth** [comment]
LGTM.  Let's also make sure to get reviews from @htuch and @fuqianggao.

File: api/envoy/admin/v3/config_dump.proto
Comment (htuch) [inline api/envoy/admin/v3/config_dump.proto:195]: For these new fields, without corresponding Envoy implementation, I'd like to mark them `[#not-implemented-yet:]`.

**Review by htuch — COMMENTED** [review]
LGTM, but would like to hear from @fuqianggao. I think this it would be good to get a plan for how the Envoy-side of the implementation will land, since this ultimately will be needed to make this generally useful.

File: api/envoy/admin/v3/config_dump.proto
Comment (lidizheng) [reply api/envoy/admin/v3/config_dump.proto:195]: Done.

File: api/envoy/admin/v3/config_dump.proto
Comment (lidizheng) [reply api/envoy/admin/v3/config_dump.proto:62]: @markdroth

Comment (markdroth) [reply api/envoy/admin/v3/config_dump.proto:62]: I don't think so.  The nonce is only relevant in the xDS stream itself, not for reporting purposes.

Comment (fuqianggao) [reply api/envoy/admin/v3/config_dump.proto:62]: I am thinking from control plane's perspective. When this is used in CSDS, it's more for debugging purpose. So might be useful in that sense?

Comment (markdroth) [reply api/envoy/admin/v3/config_dump.proto:62]: I don't think so.  A nonce is only valid within the context of an individual xDS stream.  When that stream breaks and the client establishes a new one, the previous nonce is no longer relevant.  Given that, I don't see how it could be useful for reporting purposes from either the client or the server -- it's a completely ephemeral value.

Comment (fuqianggao) [reply api/envoy/admin/v3/config_dump.proto:62]: Make sense.

**Review by fuqianggao — DISMISSED** [review]
LGTM

File: api/envoy/admin/v3/config_dump.proto
Comment (htuch) [inline api/envoy/admin/v3/config_dump.proto:62]: I think this also needs not-implemented-hide.

File: api/envoy/admin/v3/config_dump.proto
Comment (htuch) [inline api/envoy/admin/v3/config_dump.proto:141]: Nit: `*error_state*`. Same in other comments below.

**Review by htuch — CHANGES_REQUESTED** [review]
LGTM modulo final comments.
/wait

File: api/envoy/admin/v3/config_dump.proto
Comment (lidizheng) [reply api/envoy/admin/v3/config_dump.proto:63]: Done.

File: api/envoy/admin/v3/config_dump.proto
Comment (lidizheng) [reply api/envoy/admin/v3/config_dump.proto]: Done.

**Review by lidizheng — COMMENTED** [review]
@htuch Thanks for the suggestions. PTALA.

**Comment by lidizheng** [comment]
The CI result looks like a flake. Please advice...

```
2021-01-19T21:53:17.4450407Z ##[section]Starting: Bash
2021-01-19T21:53:17.4457458Z ==============================================================================
2021-01-19T21:53:17.4458296Z Task         : Bash
2021-01-19T21:53:17.4458686Z Description  : Run a Bash script on macOS, Linux, or Windows
2021-01-19T21:53:17.4459025Z Version      : 3.179.0
2021-01-19T21:53:17.4459327Z Author       : Microsoft Corporation
2021-01-19T21:53:17.4459729Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/bash
2021-01-19T21:53:17.4460175Z ==============================================================================
2021-01-19T21:53:17.8386111Z Generating script.
2021-01-19T21:53:17.8398305Z Script contents:
2021-01-19T21:53:17.8398784Z ./ci/do_ci.sh verify_examples
2021-01-19T21:53:17.8406985Z ========================== Starting Command Output ===========================
2021-01-19T21:53:17.8418007Z [command]/bin/bash --noprofile --norc /home/vsts/work/_temp/12bca6fc-0515-4538-9582-361216450100.sh
2021-01-19T21:53:17.8488853Z building using  CPUs
2021-01-19T21:53:17.8489398Z building for 
2021-01-19T21:53:17.8502423Z verify examples...
2021-01-19T21:54:16.2079297Z Error processing tar file(exit status 1): unexpected EOF
2021-01-19T21:54:16.2160140Z ##[error]Bash exited with code '1'.
2021-01-19T21:54:16.2207793Z ##[section]Finishing: Bash
```

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Comment by lidizheng** [comment]
I'm not authorized to merge. @htuch can you help to merge this PR? If there is any changes needed, I'm happy to update.

**Comment by asraa** [comment]
Re-running the failing CI test (doesn't appear to be a problem with the PR)...

**Comment by asraa** [comment]
Could you please try a main merge as well? On slack, it doesn't seem people are hitting this :/

**Comment by lidizheng** [comment]
(All checks have passed.)

**Pull request merged**
